test/common/tls/integration/ssl_integration_test.cc

```diff
@@ -867,6 +893,41 @@ TEST_P(SslCertficateIntegrationTest, ServerEcdsaClientEcdsaOnly) {
   checkStats();
 }
 
+// Server has RSA and ECDSA P-256 certificates, client only supports
+// P-384 curve. We fall back to RSA certificate.
+TEST_P(SslCertficateIntegrationTest, ServerRsaServerEcdsaP256EcdsaClientEcdsaP384Only) {
+  server_rsa_cert_ = true;
+  server_ecdsa_cert_ = true;
+  // We have to set the server curves to allow for ECDH negotiation where client
+  // doesn't support P-256
+  server_curves_.push_back("P-256");
+  server_curves_.push_back("P-384");
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection(ecdsaP384OnlyClientOptions());
+  };
+  testRouterRequestAndResponseWithBody(1024, 512, false, false, &creator);
+  const std::string counter_name = listenerStatPrefix("ssl.sigalgs.rsa_pss_rsae_sha256");
+  Stats::CounterSharedPtr counter = test_server_->counter(counter_name);
+  test_server_->waitForCounterGe(counter_name, 1);
+  EXPECT_EQ(1U, counter->value());
+  checkStats();
+}
+
+TEST_P(SslCertficateIntegrationTest, ServerRsaServerEcdsaP521EcdsaClientEcdsaP256P384Only) {
+  server_rsa_cert_ = true;
+  server_ecdsa_cert_ = true;
+  server_ecdsa_cert_name_ = "server_ecdsa_p521";
+  ConnectionCreationFunction creator = [&]() -> Network::ClientConnectionPtr {
+    return makeSslClientConnection(ecdsaP256P384OnlyClientOptions());
+  };
+  testRouterRequestAndResponseWithBody(1024, 512, false, false, &creator);
+  const std::string counter_name = listenerStatPrefix("ssl.sigalgs.rsa_pss_rsae_sha256");
+  Stats::CounterSharedPtr counter = test_server_->counter(counter_name);
+  test_server_->waitForCounterGe(counter_name, 1);
+  EXPECT_EQ(1U, counter->value());
+  checkStats();
+}
+
```

Comment:
Please add tests for successful use of P-384 and P-521 (eg client supports all curves, and server has P-384 or P-521, plus RSA.